File: typescript/examples/agents/01-agent-with-schema.ts
// Agent with structured output — use a Zod schema to get typed results
import { z } from "zod";
import { Agent } from "../../src/agent/index.js";

const SummarySchema = z.object({
  title: z.string().describe("A short title for the summary"),
  key_points: z.array(z.string()).describe("3-5 key points"),
  sentiment: z.enum(["positive", "negative", "neutral"]).describe("Overall sentiment"),
});

const agent = new Agent({
  name: "summarizer",
  instructions: "You summarize text and extract structured information.",
  outputSchema: SummarySchema,
});

const text = `
  The new product launch exceeded all expectations. Customer reviews have been
  overwhelmingly positive, with particular praise for the intuitive design and
  performance improvements. Sales figures are up 40% compared to last quarter,
  and support tickets have actually decreased despite the larger user base.
`;

const result = await agent.run(`Summarize this text: ${text}`);

// result.output is automatically typed — no cast needed!
console.log("Title:", result.output.title);
console.log("Key points:");
for (const point of result.output.key_points) {
  console.log(" -", point);
}
console.log("Sentiment:", result.output.sentiment);
console.log("Iterations:", result.meta.iterations);

File: python/examples/agents/01_agent_with_schema.py
# Agent with structured output — Pydantic model as output_schema
# The agent returns validated, typed data instead of free text.

import asyncio

from pydantic import BaseModel

from opperai.agent import Agent


class Summary(BaseModel):
    title: str
    key_points: list[str]
    sentiment: str  # "positive", "negative", or "neutral"


async def main() -> None:
    agent = Agent(
        name="summarizer",
        instructions=(
            "You summarize text into structured form. "
            "Return a title, key points, and overall sentiment (positive/negative/neutral)."
        ),
        output_schema=Summary,
    )

    result = await agent.run(
        "Python 3.12 brings major performance improvements with a new JIT compiler, "
        "better error messages, and improved typing support. The community response "
        "has been overwhelmingly positive, with many developers praising the faster "
        "execution speeds and more helpful debugging experience."
    )

    # result.output is a validated Summary instance
    summary = result.output
    print(f"Title: {summary.title}")
    print(f"Sentiment: {summary.sentiment}")
    print("Key points:")
    for point in summary.key_points:
        print(f"  - {point}")
    print(f"\nTokens: {result.meta.usage.total_tokens}")


if __name__ == "__main__":
    asyncio.run(main())
